milabench/metrics/sqlalchemy.py

```diff
@@ -0,0 +1,423 @@
+        state.step += 1
+
+    def on_phase(self, entry):
+        pass
+
+    def on_error(self, entry):
+        state = self.pack_state(entry)
+        state.error += 1
+
+    def on_line(self, entry):
+        pass
+
+    def _push_metric(
+        self,
+        run_id,
+        pack_id,
+        name,
+        value,
+        gpu_id=None,
+        job_id=None,
+        namespace=None,
+        unit=None,
+    ):
+        if not isinstance(value, numbers.Number):
+            print(f"Unexpected value {value} for metric {name}")
+            return
+
+        self.pending_metrics.append(
+            Metric(
+                exec_id=run_id,
+                pack_id=pack_id,
+                order=time.time(),
+                name=name,
+                namespace=namespace,
+                unit=unit,
+                value=value,
+                gpu_id=gpu_id,
+                job_id=job_id,
+            )
+        )
+
+    def _change_gpudata(self, run_id, pack_id, k, v, jobid):
+        for gpu_id, values in v.items():
+            for metric, value in values.items():
+                if metric == "memory":
+                    use, mx = value
+                    value = use / mx
+
+                self._push_metric(
+                    run_id, pack_id, f"gpu.{metric}", value, gpu_id, job_id=jobid
+                )
+
+    def on_data(self, entry):
+        state = self.pack_state(entry)
+
+        if state.step != DATA:
+            import sys
+
+            sys.exit(-1)
+
+        assert state.step == DATA
+
+        run_id = self.run._id
+        pack_id = state.pack._id
+        job_id, gpu_id = _get_pack_ids(state.pack)
+
+        if "progress" in entry.data:
+            return
+
+        data = deepcopy(entry.data)
+
+        # GPU
+        gpudata = data.pop("gpudata", None)
+        if gpudata is not None:
+            # GPU data would have been too hard to query
+            # so the gpu_id is moved to its own column
+            # and each metric is pushed as a separate document
+            self._change_gpudata(run_id, pack_id, "gpudata", gpudata, job_id)
+        else:
+            # Standard
+            unit = data.pop("units", None)
+            task = data.pop("task", None)
+
+            if len(data) == 1:
+                k, v = list(data.items())[0]
+
+                self._push_metric(
+                    run_id,
+                    pack_id,
+                    k,
+                    v,
+                    gpu_id=gpu_id,
+                    job_id=job_id,
+                    unit=unit,
+                    namespace=task,
+                )
+            else:
+                print(f"Unknown format {entry.data}")
+
+        if len(self.pending_metrics) >= self.batch_size:
+            self._bulk_insert()
+
+    def _bulk_insert(self):
+        if len(self.pending_metrics) <= 0:
+            return
+
+        # stmt = insert(Metric).values(self.pending_metrics)
+        # self.session.execute(stmt)
+
```

Comment:
I believe you forgot to remove this comment.

Reply:
done